event_model/__init__.py

```diff
@@ -786,19 +786,35 @@ def fill_event_page(self, doc, include=None, exclude=None):
     def fill_event(self, doc, include=None, exclude=None, inplace=None):
         try:
             filled = doc['filled']
+            from_datakeys = False
         except KeyError:
             # This document is not telling us which, if any, keys are filled.
             # Infer that none of the external data is filled.
             descriptor = self._descriptor_cache[doc['descriptor']]
             filled = {key: 'external' in val
                       for key, val in descriptor['data_keys'].items()}
+            from_datakeys = True
         for key, is_filled in filled.items():
             if exclude is not None and key in exclude:
                 continue
             if include is not None and key not in include:
                 continue
             if not is_filled:
-                datum_id = doc['data'][key]
+                try:
+                    datum_id = doc['data'][key]
+                except KeyError as err:
+                    if from_datakeys:
+                        raise MismatchedDataKeys(
+                            "The documents are not valid.  Either because they "
+                            "were recorded incorrectly in the first place, "
+                            "corrupted since, or exercising a yet-undiscovered "
+                            "bug in reader. event['data'].keys()"
+                            "must equal descriptor['data_keys'].keys()."
+                            f"event['data'].keys(): {doc['data'].keys()}, "
+                            "descriptor['data_keys'].keys(): "
+                            f"{self._descriptor_cache()}") from err
+                    else:
+                        raise err
```

Comment:
This case would mean that there are keys in `event['filled']` which are not in `event['data']`, which is another way in which the keys could be mismatched. Is it worth raising the same exception with different message text here? We haven't seen this failure mode yet, and I doubt we'd see it from _databroker_ (because Events at rest have no `'filled'` key) but it might come up in stream applications with Events over the wire, which can have a `'filled'` key.